Sendpoi

gps1539 opened the pull request:
Added geocode lookup from OSM to get lat, lon and name

swimsystems commented on bimmer_connected/cli.py:
```diff
@@ -6,6 +6,8 @@
 import json
 import os
 import time
+import urllib.request
```
I'd suggest to use `requests` for http calls to the server.

gerard33 replied:
Agreed, `requests` is used in other parts of the library as well.

gps1539 replied:
I can change my code to use requests, if that is preferred. Probably be a couple of days till I can make the changes and test again.

gps1539 replied:
I found some time this evening and changed to use requests. I tested and send a few POIs to my car successfully, but good if others can test. 

I git pushed my changes and submitted a new PR. Apologies in advance if I did not follow the right process. Please let me know.

swimsystems commented on bimmer_connected/cli.py:
```diff
@@ -117,7 +118,33 @@ def image(args) -> None:
     vehicle = account.get_vehicle(args.vin)
-    poi = PointOfInterest(args.latitude, args.longitude, name=args.name)
+    address = (str(' '.join(args.address)).replace(' ', '+'))
+    url = "https://nominatim.openstreetmap.org/?format=json&addressdetails=1&q="+address+"&format=json&limit=1"
```
See https://realpython.com/python-requests/#query-string-parameters. I think this is much easier than +ing all arguments together.

Also imho it would make sense to split this code into a separate function for example `address_to_poi`. That way it could be reused more easily and is easier to test.

rikroe replied:
I really like the idea of a separate function! So instead of calling `bimmerconnected sendpoi ...` maybe we could call `bimmerconnected sendpoi_from_adress ...` (which then of course can use the existing `send_poi` function.

gerard33 replied:
Me too. Can you update this? Then we can do some final testing and merge the SendPOI PRs.

rikroe commented:
Hey, sorry for not really looking into this PR. Like I wrote before, many thanks for trying out and prototyping this implementation. 

I think for this PR to be submitted it would be great if you could implement some changes:
- It is not possible anymore to make a cli call with `lat`, `lon` or any of the singular address parts from #103. Please create a **second command line option** (for example `sendpoi_from_address`) to keep the existing cli argument working.
- I was expecting to add a query string (i.e. `--address "Street 1, City"`) but your implementation expects a list `--address Street 1, City` that needs then to be joined together. Using `' '.join()` on a string will add a whitespace between every single letter. I would prefer to **enter the address using quotes** as this also would be in line with command line options dealing with whitespaces.
- When a value is not found in the OSM response, you set it to `'null'`, i.e. a string. **Please set it to** `None` so that the respective values won't be transmitted to the car at all. If you want to shorten your code, e.g. `street = g.get("address", {}).get("road")` would use python `dict`'s `get` method, returning `None` if the respective key is not found.

While the above points are a must for me, please also think about the following suggestions:
- When using requests, as mentioned by @swimsystems, the `params` attribute makes it much clearer to stitch the url together. In the current implementation there is also a doubled URL parameter. Working example: `requests.get(url="https://nominatim.openstreetmap.org/", params={"format": "json", "addressdetails": 1, "q": address, "limit": 1}).json()[0]`

It would be great to have your idea included, especially as the two other PRs are now merged!

Thanks!

gps1539 commented:
Thanks, I will be traveling for work over the next 2 weeks and it will be hard to look at this. I'm happy if someone else has time to make the Must changes.

The --address argument should accept quotes or no quotes, i.e. "123 main st, city" or 123 main st, city. I can double check when I get back, but I'm 99% sure it does.

gps1539 commented:
Back from my trip and I will find time to work on this in the next week.

gps1539 commented:
Hi 
I made updates base on the feedback above. 

1. I added back the sendpoi parser and function and changed my code to sendpoi_from_address.
I've tested sending poi from addresses to my car and it is working well. I also tested using sendpoi and providing lat and lon and that seems to be working as well.
2. You could enter an address in quotes or not, both work. i.e. 
-a  "114520 Camden Ave San Jose"
-a 114520 Camden Ave San Jose
both send the poi to my car
3. I also change 'null' to None when street etc. is not found. No more 'null' being sent to my car :-)

4. The url I'm using to call OSM is based on their guidance and seems to work well

rikroe commented:
Cool, many thanks for the updates!

> 2. You could enter an address in quotes or not, both work. i.e.
>    -a  "114520 Camden Ave San Jose"
>    -a 114520 Camden Ave San Jose
>    both send the poi to my car

This does not work for me with Python 3.6 on Windows 10 or Ubuntu (I've added `print(type(args.address), args.address)` to get some quick debug output):
```
(bimmer_connected) PS > python -m bimmer_connected.cli sendpoi_from_address EMAIL PW rest_of_world VIN -a 'Zoo Frankfurt'      
win32  -  3.6.8 (tags/v3.6.8:3c6b436a57, Dec 24 2018, 00:16:47) [MSC v.1916 64 bit (AMD64)]
<class 'str'> Zoo Frankfurt
<class 'str'> Z+o+o+++F+r+a+n+k+f+u+r+t

(bimmer_connected) PS > python -m bimmer_connected.cli sendpoi_from_address EMAIL PW rest_of_world VIN -a Zoo Frankfurt        
usage: cli.py [-h]
              {status,fingerprint,lightflash,image,sendpoi,sendpoi_from_address}
              ...
cli.py: error: unrecognized arguments: Frankfurt
```

I'd really rather go with the `request` `params` implementation, then you wouldn't have to replace spaces with `+` and all others sorts of conversions would be taken care of: 
```
g = requests.get(
    "https://nominatim.openstreetmap.org",
    params={
        "q": args.address,
        "format": "json",
        "addressdetails": 1,
        "limit": 1,
    },
).json()[0]
```

gps1539 commented:
Okay, let me try that approach. I'm running arch linux and python 3.8, which may explain why it works for me.

gps1539 commented:
I made the change and ran a couple of tests. It works with quotes (single & double), but without quotes can now get the wrong results such as 
-a accrington, united kingdom
'43.1139558 -76.3960453 test None Town of Van Buren 13112 United States'

I will see if I can resolve the above with a request params implementation, but good if someone can test with an older version of python.

rikroe commented:
Thanks, will test with python 3.6 tomorrow (and probably some others as well).

Regarding the non-quote issue: I suspect the type of `args.address` is `list`. So maybe try the following:
```
        "q": str(args.address),
```
This should convert a list to a string and if a string comes in, do nothing. 
Hope it helps!

gps1539 commented:
Thanks @rikroe, your suggestion triggered an idea. Always making args.address a list solved that issue with no quotes. I just pushed that change.

rikroe commented:
Well, I'm at a loss. For the love of god cannot reproduce my orignal issue either on Windows or Ubuntu (WSL) with Python 3.6.8 or 3.7.5. So I guess we're good. 

Your code works for me now, sorry for the nitpicking which I'm not able to reproduce!

gps1539 commented:
@rikroe, are you trying to repo on the old version of my code or after I made changes to try a support python 3.6?

rikroe commented:
I was trying it on your new code from [018313d](https://github.com/bimmerconnected/bimmer_connected/pull/115/commits/018313db0b8a1be0ad8c0e19ebf1f4452d026196).

But also I wasn't able to reproduce my issues with earlier code snippets - again doesn't matter, it works now :)

rikroe commented on bimmer_connected/cli.py:
```diff
@@ -117,7 +132,40 @@ def image(args) -> None:
+        g = requests.get("https://nominatim.openstreetmap.org", params={"q": address, "format": "json","addressdetails": 1,"limit": 1},).json()[0]
+    except IndexError:
+        print('\nAddress not found')
+        sys.exit(1)
```
Just realized, you should be able to call `exit(1)` without explicilty calling it from `sys`, therefore also be able to remove the import up top.

gps1539 commented:
From what I recall sys.exit is the recommended way to exit in python. This thread implies the same
https://stackoverflow.com/questions/6501121/difference-between-exit-and-sys-exit-in-python

rikroe commented:
Thanks for that piece of information! Then we're all good!

gerard33 commented:
Yeah I will have a look at this PR asap. Will also check and fix the branch conflicts.

rikroe commented:
Shouldn't this be merged since the previous version?

gerard33 commented:
I have merged SendPOI PR #131, this one can be merged as well, but first the conflicts must be resolved. Haven't done that yet in for 0.7.0 I only added the PRs #148 and #151.

gerard33 commented:
Closing as #164 replaces this functionality.
